### DISCUSSION TITLE
How to handle state of an external neotree source

### DISCUSSION THREAD

**Discussion by ghostbuster91** [nvim-neo-tree/neo-tree.nvim#1307] [Internal Development] [answered]
Hi,

I finally started working on a new source for neotree. 
I was able to come up with something that works however I have a problem with keeping the state of my source (i.e. which items are collapsed/expanded). 

Whenever I move focus to another window my neotree source gets reloaded due to `navigate` method being called and all information about expanded nodes is lost. I tried reading existing sources, but didn't find anything which seemed to be relevant at the first glance. 

Any idea what could that be?

I will share my full source code soon, I just want to clean it up a bit before.

**Comment by cseickel** [comment]
The trick is to create a table on your state called `default_expaned_nodes` and put the node IDs in that array. There is also an `item._is_expanded` property that is read when the items are being rendered although I don't see any built-in sources actually setting that proeprty.

**Reply by ghostbuster91** [reply, marked as answer]
I was actually able to do it without using `default_expanded_nodes`. Maybe we have not understood each other. I admit, my problem description was quite shallow, but the same applied to my understanding of the problem at that time. 

Now I know a little bit better. There are different cases and each one has to be handled differently. 
The case that I described, switching windows is the easiest. If we already have the tree and the window is visible then there is nothing to be done. My problem was that I was trying to do in this case.

Then I ran into another case which is when you are switching between different sources in neotree (e.g. by calling `:Neotree buffers`, `:Neotree my_source`). I tried calling `renderer.redraw` but it seemed to take no effect. The key turned out to be `renderer.acquire_window`. 

I think I read most of the documentation in the repository yet I haven't found anything about the the necessity to call these functions in different cases when implementing external source. I think, that we should document some kind of protocol that external source should obey. This will be extremely helpful for other people to integrate with neotree. 

I can start it, since I am already working on a new source :)

**Reply by cseickel** [reply]
Documentation could definitely be a lot better. Any help along those lines is greatly appreciated.

**Comment by miversen33** [comment]
Feel free to borrow/beg/steal code from [netman.nvim](https://github.com/miversen33/netman.nvim/tree/v1.15/lua/netman/ui/neo-tree) I went down this road already. You unfortunately need to serialize the entire tree and keep track of it yourself

**Reply by ghostbuster91** [reply]
How come? I was actually able to achieve what I want without keeping track about the tree myself. Neotree passes the `extra` parameter during the nui-node creation https://github.com/nvim-neo-tree/neo-tree.nvim/blob/main/lua/neo-tree/ui/renderer.lua#L272 so this is all I need. Maybe your use-case was more complex or maybe I am yet to discover the need for this approach but I hope not :)

**Reply by miversen33** [reply]
I don't actually remember why I ended up having to serialize my tree and track it myself. I do recall that that is how I ended up solving [netman #154](https://github.com/miversen33/netman.nvim/issues/154). 

I just remember that I ended up having to track the tree state myself and shim creation of the tree within neo-tree to get what I wanted. That being the state not being created each time the source was reopened.

I'm glad you got it working for you though!

**Comment by ghostbuster91** [comment]
Thanks for the suggestions. `default_expanded_nodes` looks promising but I think there should be another way. Does the `document_symbol` source reloads every time you change the focus? I guess not. I will take a closer look at it, maybe I will spot something.
